Consolidate retrieving state when resuming partially read JSON

bastbu opened the pull request:
This PR consolidates the duplication around resuming when partially reading JSON.

atifaziz reviewed (commented):
This is putting essentially an `if`-`else` or ternary conditional operator behind a method and a delegate, which has a few effects:

- The argument validation adds an additional branch.
- The `throw` means it won't get in-lined by JIT (this be avoided by makes code messy).
- The delegate invocation introduces a virtual call.
- It's too easy to introduce a closure (requires care to sprinkle `static` at all call sites that can't be enforced by the method).

Do you reckon we're consolidating much here?

bastbu commented:
Could be. When I was writing #27, I was thrown off by always having to write the same logic for resumption - check if it's resuming, pop, cast otherwise default.

I was also wondering why the `Utf8JsonReader` exposes the `Push` and `Pop` and hence I thought that moving the `ResumeOrDefault` as the counterpart of the `Suspend` to the `Utf8JsonReader` makes sense.

To address both of our concerns, would a `bool TryResume<T>([NotNullWhen(true)] out T? state)` make more sense for more consolidation, encapsulation (`Pop` and `Push` can be made `private`) while addressing your concerns at the same time (sketched in 5b10e42fc508ccbfc8aa03a924e710c2a508186a)?

atifaziz reviewed (approved):
Much better with the follow-up! Thanks!
